libnet/src/ioctl.rs

```diff
@@ -1519,19 +1520,43 @@ pub(crate) fn create_vnic(
         let mut arg = VnicIocCreate {
             link_id,
             vnic_id: id,
-            mac_addr_type: VnicMacAddrType::Auto,
-            mac_len: 0,
-            mac_prefix_len: 3,
+            mac_addr_type: match mac {
+                None => VnicMacAddrType::Auto,
+                Some(_) => VnicMacAddrType::Fixed,
+            },
+            mac_len: match &mac {
+                None => 0,
+                Some(mac) => mac.len() as u32,
+            },
+            mac_prefix_len: match mac {
+                None => 3,
+                Some(_) => 0,
+            },
             mac_slot: -1,
             vid: 0,
             vrid: 0,
             af: AF_UNSPEC as i32,
             flags: 0,
             ..Default::default()
         };
-        arg.mac_addr[0] = 0x02;
-        arg.mac_addr[1] = 0x08;
-        arg.mac_addr[2] = 0x20;
+        match mac {
+            None => {
+                // This seems to be the OUI prefix used for bhyve/virtio NICs
+                // but I cannot find a documentation reference that actually
+                // says so.
+                arg.mac_addr[0] = 0x02;
+                arg.mac_addr[1] = 0x08;
+                arg.mac_addr[2] = 0x20;
+            }
+            Some(mac) => {
+                if mac.len() > 20 {
+                    return Err(Error::BadArgument("mac too long".to_string()));
+                }
```

Comment:
What happens if the mac is too short?

Reply:
The buffer would be filled with whatever number of bytes were given followed by zeros. The `mac_len` would be set to whatever was given, and if it's not something sane, like 6 for ethernet, I suspect the ioctl will fail.